### PR TITLE
[MD friendly] Prevent bans for participant arrays (providing more than 1 id)

### LINKED ISSUE #706: Number banned when remove 2 or more numbers from group
### Bug description
I tried to remove 2 participants from a group in a row and i got banned from WhatsApp.
I don't know if i just have "lucky" and got banned or it's really a bug.

### Reproduction steps
1. Call api node with params
image

2. Numbers removed from group and then i got banned

### Relevant code
image

### Environment (please complete the following information):
**WhatsApp**
 - Account type: STANDARD
 - Device OS: ANDROID

**Library**
 - NodeJS version: 14.16.0
 - whatsapp-web.js version: 1.12.6

### LINKED ISSUE #1000: Is there anyway to use the addparticipants function without getting banned ?
Whenever i use this function some times my number is banned

### LINKED ISSUE #1880: addParticipants is banning the account
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When trying to add a contact to a group that I am an administrator, whatsapp bans my number.

### Expected behavior

I would like to add the number and not get banned.

### Steps to Reproduce the Bug or Issue

**1 -** Add the phone number to the contact list
**2 -** Choose which group you want to add the contact to
**3 -** Find the ID of the contact I want to add
**4 -** I add the contact in the group selected in item 2.
**5 -** I'm banned

### Relevant Code

```
client.on('ready', () => {
    console.log('Client is ready!');

    client.getChats().then((chats) => {
        const myGroupName = 'MY GROUP';
        const myGroup = chats.find((chat) => chat.name === myGroupName);
        client.getContacts().then((contacts) => {
            const contactToAdd = contacts.find(
                // Finding the contact Id using the contact's name
                (contact) => contact.name === 'Paulo'
              );
            
            if (contactToAdd) {
                console.log(' :D ');
                myGroup.addParticipants([contactToAdd.id._serialized]);
            }else{
                console.log(' :( ');
            }
        });
    })
});
```

### Browser Type

Google Chrome

### WhatsApp Account Type

Standard

### Does your WhatsApp account have multidevice enabled?

Yes, I am using Multi Device

### Environment

**OS:** Ubuntu 20.04
**Phone:** Android
**whatsapp-web.js:** Last Version 
**WhatsApp Web Version:** 2.2247.6
**Node:** V.19.2

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by purpshell**
you can now use the participant array property in the method without getting banned for providing more than one entry. 

Also added sleep timer for adding a delay in milliseconds between actions (added Foo, wait 500ms, added Bar)

### CODE REVIEW

**Comment by PurpShell** [comment]
> Can you explain why this way won't get banned?

The function was getting flagged because of us passing all the ids in in one fell swoop, now we iterate each id and perform the action, giving whatsapp only 1 number

**Comment by PurpShell** [comment]
> > Can you explain why this way won't get banned?
> 
> The function was getting flagged because of us passing all the ids in in one fell swoop, now we iterate each id and perform the action, giving whatsapp only 1 number 

It is like doing a for loop with an array with one item like before, now we just simply do that loop internally

**Comment by tuyuribr** [comment]
Sounds good, maybe a sleep too to avoid too fast bans?

**Comment by shirser121** [comment]
addParticipants return this error
```js
Uncaught TypeError: t.map is not a function
```

File: src/structures/GroupChat.js [hunk 1]
```diff
@@ -55,54 +56,86 @@ class GroupChat extends Chat {
...
-        return await this.client.pupPage.evaluate((chatId, participantIds) => {
+    async addParticipants(participantIds, sleep = null) {
+        return await this.client.pupPage.evaluate(async (chatId, participantIds, sleep) => {
             const chatWid = window.Store.WidFactory.createWid(chatId);
             const participantWids = participantIds.map(p => window.Store.WidFactory.createWid(p));
-            return window.Store.GroupParticipants.sendAddParticipants(chatWid, participantWids);
-        }, this.id._serialized, participantIds);
+            const status = [];
+            for (const participantWid of participantWids) {
+                status.push(await window.Store.GroupParticipants.sendAddParticipants(chatWid, participantWid));
```
Comment (unilogica) [inline src/structures/GroupChat.js:69]: status.push(await window.Store.GroupParticipants.sendAddParticipants(chatWid, [participantWid]));

File: src/structures/GroupChat.js [hunk 2]
```diff
@@ -55,54 +56,86 @@ class GroupChat extends Chat {
...
-        return await this.client.pupPage.evaluate((chatId, participantIds) => {
+    async removeParticipants(participantIds, sleep = null) {
+        return await this.client.pupPage.evaluate(async (chatId, participantIds, sleep) => {
             const chatWid = window.Store.WidFactory.createWid(chatId);
             const participantWids = participantIds.map(p => window.Store.WidFactory.createWid(p));
-            return window.Store.GroupParticipants.sendRemoveParticipants(chatWid, participantWids);
-        }, this.id._serialized, participantIds);
+            const status = [];
+            for (const participantWid of participantWids) {
+                status.push(await window.Store.GroupParticipants.sendRemoveParticipants(chatWid, participantWid));
```
Comment (unilogica) [inline src/structures/GroupChat.js:90]: status.push(await window.Store.GroupParticipants.sendRemoveParticipants(chatWid, [participantWid]));

File: src/structures/GroupChat.js [hunk 3]
```diff
@@ -55,54 +56,86 @@ class GroupChat extends Chat {
...
-        return await this.client.pupPage.evaluate((chatId, participantIds) => {
+    async promoteParticipants(participantIds, sleep = null) {
+        return await this.client.pupPage.evaluate(async (chatId, participantIds, sleep) => {
             const chatWid = window.Store.WidFactory.createWid(chatId);
             const participantWids = participantIds.map(p => window.Store.WidFactory.createWid(p));
-            return window.Store.GroupParticipants.sendPromoteParticipants(chatWid, participantWids);
-        }, this.id._serialized, participantIds);
+            const status = [];
+            for (const participantWid of participantWids) {
+                status.push(await window.Store.GroupParticipants.sendPromoteParticipants(chatWid, participantWid));
```
Comment (unilogica) [inline src/structures/GroupChat.js:111]: status.push(await window.Store.GroupParticipants.sendPromoteParticipants(chatWid, [participantWid]));

File: src/structures/GroupChat.js [hunk 4]
```diff
@@ -55,54 +56,86 @@ class GroupChat extends Chat {
...
-        return await this.client.pupPage.evaluate((chatId, participantIds) => {
+    async demoteParticipants(participantIds, sleep = null) {
+        return await this.client.pupPage.evaluate(async (chatId, participantIds, sleep) => {
             const chatWid = window.Store.WidFactory.createWid(chatId);
             const participantWids = participantIds.map(p => window.Store.WidFactory.createWid(p));
-            return window.Store.GroupParticipants.sendDemoteParticipants(chatWid, participantWids);
-        }, this.id._serialized, participantIds);
+            const status = [];
+            for (const participantWid of participantWids) {
+                status.push(await window.Store.GroupParticipants.sendDemoteParticipants(chatWid, participantWid));
```
Comment (unilogica) [inline src/structures/GroupChat.js:132]: status.push(await window.Store.GroupParticipants.sendDemoteParticipants(chatWid, [participantWid]));

File: src/util/Util.js [hunk 5]
```diff
@@ -181,6 +181,10 @@ class Util {
     static setFfmpegPath(path) {
         ffmpeg.setFfmpegPath(path);
     }
+    
+    static sleep(ms) {
+        return new Promise(resolve => setTimeout(resolve, ms));
```
Comment (unilogica) [inline src/util/Util.js:186]: I rewrite Promises directly in my GroupChat (for some reason) Error: Evaluation failed: ReferenceError: Util is not defined with `const Util = require('../util/Util');`

**Review by unilogica — CHANGES_REQUESTED** [review]
I understand the proposed logic, and it makes sense to get banned for bulk actions. but when applying this PR and testing it I see that isn't promoting admins. 

So as it sends each promote request at once, I got in the current WhatsApp version code expects an array, so it wraps it back inside an array on the `for` and it worked for me - and I feel thats happening because Client.createGroup waits for an array (When you create a Group, it expects an array of partcicipants and in WhatsApp you can create a Group With some participants / than You will be instabanned When you to exceed the limit of App).

**Comment by PurpShell** [comment]
everything should be fixed, well that was some spaghetti code hahahh

**Comment by PurpShell** [comment]
> I understand the proposed logic, and it makes sense to get banned for bulk actions. but when applying this PR and testing it I see that isn't promoting admins. 
> 
> So as it sends each promote request at once, I got in the current WhatsApp version code expects an array, so it wraps it back inside an array on the `for` and it worked for me - and I feel thats happening because Client.createGroup waits for an array (When you create a Group, it expects an array of partcicipants and in WhatsApp you can create a Group With some participants / than You will be instabanned When you to exceed the limit of App).

This is the way the app itself implements the adding/removal..

Anyways not wrapping the id in an array was a typo from my end 🙃

File: src/structures/GroupChat.js [hunk 6]
```diff
@@ -52,57 +53,65 @@ class GroupChat extends Chat {
...
+     * @param {?number} sleep optional, amount to sleep in milliseconds before removing the next participant
+     * @returns {Promise<Array<Object>>}
      */
-    async removeParticipants(participantIds) {
-        return await this.client.pupPage.evaluate((chatId, participantIds) => {
-            const chatWid = window.Store.WidFactory.createWid(chatId);
-            const participantWids = participantIds.map(p => window.Store.WidFactory.createWid(p));
-            return window.Store.GroupParticipants.sendRemoveParticipants(chatWid, participantWids);
-        }, this.id._serialized, participantIds);
+    async removeParticipants(participantIds, sleep = null) {
```
Comment (tuyuribr) [inline src/structures/GroupChat.js:93]: Isn't better to have a sleep value by default?

**Review by tuyuribr — CHANGES_REQUESTED** [review]
sleep default value

File: src/structures/GroupChat.js (hunk 6 above)
Comment (PurpShell) [reply src/structures/GroupChat.js:93]: > Isn't better to have a sleep value by default?

ok, around 100ms will do

**Comment by PurpShell** [comment]
@tuyuribr done

**Review by tuyuribr — APPROVED** [review]

**Pull request closed without merging**